pkg/alertmanager/amcfg.go

```diff
@@ -298,9 +299,16 @@ func (cb *configBuilder) enforceNamespaceForInhibitRule(ir *inhibitRule, namespa
 
 // enforceNamespaceForRoute modifies the route configuration to match alerts
 // originating only from the given namespace.
-func (cb *configBuilder) enforceNamespaceForRoute(r *route, namespace string) *route {
+func (cb *configBuilder) enforceNamespaceForRoute(r *route, namespace string, enabled bool) *route {
```

Comment:
The same thing would be needed for inhibition rules. Rather than adding a new bool parameter to all `enforceNamespaceFor...()` methods, it might be easier to refactor configBuilder and abstracting away the (lack of) enforcement logic into a separate component.

```golang
type enforcer interface {
        processRoute(*route) *route
        processInhibitRule(*inhibitRule) *inhibitRule
}

// No enforcement
type noopEnforcer struct{}
func (ne *noopEnforce) processRoute(r *route) *route { return r }
func (ne *noopEnforce) processInhibitRule(ir *inhibitRule) *inhibitRule { return ir }

// Enforcing the namespace label
type namespaceEnforcer struct { namespace string }
func (ne *namespaceEnforcer) processRoute(r *route) *route {
  // similar to enforceNamespaceForInhibitRule
}
func (ne *namespaceEnforcer) processInhibitRule(ir *inhibitRule) *inhibitRule {
  // similar to enforceNamespaceForRoute
}

type configBuilder struct {
        ...
        enforcer  enforcer
}

func (cb *configBuilder) convertRoute(in *monitoringv1alpha1.Route, crKey types.NamespacedName) *route {
   ...
   return cb.enforcer.processRoute(ir)
}

func (cb *configBuilder) convertInhibitRule(in *monitoringv1alpha1.InhibitRule) *inhibitRule {
    ...
    return cb.enforcer.processInhibitRule(ir)
}
```

Reply:
This pattern looks quite nice, I modified it so e.g. `cb.enforcer.processRoute` will no be called on the returned route in `convertRoute` but like this:
```
subRoutes = append(subRoutes,
    cb.enforcer.processRoute(
        cb.convertRoute(
	    amConfigs[amConfigIdentifier].Spec.Route,
	    crKey,
	),
    ),
)
```
This ensures that on routes from a global config (so the current behaviour is kept) and on child routes (seems to be obsolete) no route processing will be applied.